streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java

```diff
@@ -104,7 +110,11 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
             final KTableSource<K, V> ktableSource = (KTableSource<K, V>) processorParameters.processorSupplier();
             if (ktableSource.queryableName() != null) {
                 topologyBuilder.addStateStore(storeBuilder, nodeName());
-                topologyBuilder.markSourceStoreAndTopic(storeBuilder, topicName);
+
+                if (shouldReuseSourceTopicForChangelog) {
```

Comment:
This is the optimization and we only apply it for materialized `KTable` objects.